lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventTrackerImpl.java

```diff
@@ -0,0 +1,187 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.RangeSet;
+import com.google.common.collect.TreeRangeSet;
+import com.google.common.primitives.Ints;
+import com.palantir.lock.LockDescriptor;
+
+/**
+ * Note on concurrency:
+ *
+ * Multiple threads may simultaneously try to update the tracker, potentially with overlapping updates. At the same
+ * time, it is allowed to return a slightly stale version on calls to currentState, so we do not want to block on
+ * updates. This is implemented as follows.
+ *
+ * 1. Updates are synchronized, and the values of watches, singleLocks, openLocksEvents, lastKnownVersion, and leaderId
+ *    are only changed within the synchronized block. This ensures that no two threads are going to conflict during an
+ *    update, or do unnecessary work.
+ * 2. Furthermore, access to watches, singleLocks, openLocksEvents, and lastKnownVersion is generally guarded by a
+ *    read/write lock (except for reads in the synchronized block, which are safe due to 1.). This allows calls to
+ *    currentState() to return a consistent version of lock watch state using a read lock without blocking on an
+ *    in-progress update, as long as the write lock is not locked
+ */
+@ThreadSafe
+public class LockWatchEventTrackerImpl implements LockWatchEventTracker {
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private final AtomicReference<RangeSet<LockDescriptor>> watches = new AtomicReference<>(TreeRangeSet.create());
+    private final AtomicReference<Map<LockDescriptor, LockWatchInfo>> singleLocks = new AtomicReference<>(
+            ImmutableMap.of());
+    private final ConcurrentSkipListSet<UUID> openLocksEvents = new ConcurrentSkipListSet<>();
+    private final ConcurrentMap<Long, VersionedLockWatchState> startTsToLockWatchState = Maps.newConcurrentMap();
+    private volatile OptionalLong lastKnownVersion = OptionalLong.empty();
+    private volatile UUID leaderId = UUID.randomUUID();
```

Comment:
I'm not entirely sure these actually need to be volatile (since readers take either the read lock or the object lock, and the release of the write or object locks after a write happens-before the read lock or object lock respectively can be acquired). I'm not 100% convinced that removing this is safe, so maybe we should leave as is